src/meta/processors/admin/Balancer.cpp

```diff
@@ -270,82 +281,136 @@ Balancer::genTasks(GraphSpaceID spaceId,
     for (auto& h : hostDel) {
         auto& lostParts = hostParts[h];
         for (auto& partId : lostParts) {
+            LOG(INFO) << "Try balance part " << partId << " for lost host " << h;
             // check whether any peers which is alive
             auto alive = checkReplica(hostParts, activeHosts, spaceReplica, partId);
             if (!alive.ok()) {
-                LOG(ERROR) << "Error: " << alive;
+                LOG(ERROR) << "Check Replica failed: " << alive
+                           << " Part: " << partId;
                 return cpp2::ErrorCode::E_NO_VALID_HOST;
             }
             // find a host with minimum parts which doesn't have this part
             auto ret = hostWithMinimalParts(newHostParts, partId);
             if (!ret.ok()) {
-                LOG(ERROR) << "Error: " << ret.status();
+                LOG(ERROR) << "Can't find a host which doesn't have part:" << partId;
                 return cpp2::ErrorCode::E_NO_VALID_HOST;
             }
-            auto& luckyHost = ret.value();
-            newHostParts[luckyHost].emplace_back(partId);
-            tasks.emplace_back(plan_->id_,
-                               spaceId,
-                               partId,
-                               h,
-                               luckyHost,
-                               kv_,
-                               client_);
+
+            auto& targetHosts = ret.value();
+            for (auto& target : targetHosts) {
+                LOG(INFO) << "Survey target host " << target;
+                // If this space rely on group and zone we should check whether the zone
+                // which luck host is belong to have hold the same partition.
+                if (dependentOnGroup && !checkZoneLegal(h, target, partId)) {
```

Comment:
I think it is better to write another `hostWithMinimalParts` with zone instead of reuse previous code to pick one and check it.